File: transferOut/diameterOfBinaryTree.js
//diameter of Binary tree
var diameterOfBinaryTree = function (root) {
    let max = 0;
   dfs(root);
 
    return max;

    function dfs(node) {
        if (!node) return -1; // no path = lacking 1 path

    
        const left = dfs(node.left) + 1;
     
        const right = dfs(node.right) + 1;
        max = Math.max(max, left + right); // left + right => local max
        return Math.max(left, right);
    }
};

// create binary tree 
function toBinary(nums){
    if(!nums)return{};
   return segment(nums.length - 1);
    function segment(hi, lo = 0, node = {}){
        if(lo>hi)return null;
        const mid = lo + (( hi - lo) >> 1);
        node.val= nums[mid];
        node.left = segment(mid - 1, lo, {});
        node.right= segment(hi, mid + 1, {});
        return node; 
    }
}



let iter = 6;
const arr = [];
while (iter--) arr[5 - iter] = 5 - iter;


let root = toBinary(arr);
root
let diameter = diameterOfBinaryTree(root)

diameter

File: transferOut/diameterOfBinaryTree.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;
struct node
{
  int val;
  node *right = NULL;
  node *left = NULL;
};

struct Solution
{
  int diameter = 0;

  int diameterTree(node *obj)
  {
    explore(obj);
    return diameter;
  }

  int explore(node *obj)
  {
    if (!obj)
      return -1;
    int left = explore(obj->left) + 1;
    int right = explore(obj->right) + 1;
    diameter = max(diameter, left + right);
    return max(left, right);
  }

  node *toBinary(vector<int> &nums)
  {
    if (!nums.size())
      return NULL;
    return expandObject(nums, nums.size() - 1);
  }

  node *expandObject(vector<int> nums, int hi, int lo = 0, node *obj = new node())
  {
    if (lo > hi)
      return NULL;
    int mid = lo + ((hi - lo) >> 1);

    obj->val = nums[mid];

    obj->left = expandObject(nums, mid - 1, lo, new node());
    obj->right = expandObject(nums, hi, mid + 1, new node());
    return obj;
  }
};

int main()
{

  Solution *path = new Solution();
  int i = 51;
  vector<int> arr;
  while (i--)
    arr.push_back(50 - i);

  node *tree = path->toBinary(arr);

  int diameter = path->diameterTree(tree);

  cout << diameter;
  return 0;
};